packages/vite/src/node/server/middlewares/proxy.ts

```diff
@@ -78,40 +87,101 @@ export function proxyMiddleware(
   }
 
   // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
-  return function viteProxyMiddleware(req, res, next) {
+  return async function viteProxyMiddleware(req, res, next) {
     const url = req.url!
     for (const context in proxies) {
       if (
         (context.startsWith('^') && new RegExp(context).test(url)) ||
         url.startsWith(context)
       ) {
         const [proxy, opts] = proxies[context]
-        const options: HttpProxy.ServerOptions = {}
 
         if (opts.bypass) {
           const bypassResult = opts.bypass(req, res, opts)
           if (typeof bypassResult === 'string') {
             req.url = bypassResult
             debug(`bypass: ${req.url} -> ${bypassResult}`)
             return next()
-          } else if (typeof bypassResult === 'object') {
-            Object.assign(options, bypassResult)
-            debug(`bypass: ${req.url} use modified options: %O`, options)
-            return next()
           } else if (bypassResult === false) {
             debug(`bypass: ${req.url} -> 404`)
             return res.end(404)
           }
         }
 
+        const activeProxyOptions = await prepareProxyRequest(req, opts)
         debug(`${req.url} -> ${opts.target || opts.forward}`)
-        if (opts.rewrite) {
-          req.url = opts.rewrite(req.url!)
-        }
-        proxy.web(req, res, options)
+        proxy.web(req, res, activeProxyOptions)
         return
       }
     }
     next()
   }
 }
+
+async function prepareProxyRequest(
+  req: http.IncomingMessage,
+  opts: ProxyOptions
+) {
+  // req.url = req.originalUrl || req.url
+  const newProxyOptions = Object.assign({}, opts)
+
+  if (opts.router) {
+    const newTarget = await getTarget(req, opts)
+    if (newTarget) {
+      debug('[proxy] Router new target: %s -> "%s"', opts.target, newTarget)
+      newProxyOptions.target = newTarget
+    }
+  }
+  if (opts.rewrite) {
+    req.url = opts.rewrite(req.url!)
+  }
+
+  return newProxyOptions
+}
+
+async function getTarget(req: http.IncomingMessage, config: ProxyOptions) {
+  let newTarget
```

Comment:
**suggestion:** Could you please add a typedef to this `let newTarget`?
Cause it cannot inferred out of the box.

_Note to me and @patak-js:_ I could lookup if there is a eslint rule to enforce this :thinking: